crates/wasi/src/preview2/pipe.rs

```diff
@@ -210,84 +222,190 @@ impl HostInputStream for AsyncReadStream {
 }
 
 #[derive(Debug)]
-enum WriteState {
-    Ready,
-    Pending,
-    Err(std::io::Error),
+enum WriteMessage {
+    Write(Bytes),
+    Flush,
 }
 
-/// Provides a [`HostOutputStream`] impl from a [`tokio::io::AsyncWrite`] impl
-pub struct AsyncWriteStream {
-    state: Option<WriteState>,
-    sender: tokio::sync::mpsc::Sender<Bytes>,
-    result_receiver: tokio::sync::mpsc::Receiver<Result<StreamState, std::io::Error>>,
-    join_handle: tokio::task::JoinHandle<()>,
+#[derive(Debug)]
+enum FlushState {
+    Enqueued,
+    InProgress,
+    Done,
 }
 
-impl AsyncWriteStream {
-    /// Create a [`AsyncWriteStream`]. In order to use the [`HostOutputStream`] impl
-    /// provided by this struct, the argument must impl [`tokio::io::AsyncWrite`].
-    pub fn new<T: tokio::io::AsyncWrite + Send + Sync + Unpin + 'static>(mut writer: T) -> Self {
-        let (sender, mut receiver) = tokio::sync::mpsc::channel::<Bytes>(1);
-        let (result_sender, result_receiver) = tokio::sync::mpsc::channel(1);
+#[derive(Debug)]
+struct WorkerState {
+    alive: bool,
+    items: std::collections::VecDeque<WriteMessage>,
```

Comment:
This part about all the queueing makes me a bit uneasy. I'd naively expect at-most-one thing here contractually and then we'd add vectors/etc as optimizations in the future maybe.

I think this stems though from the contract of flush/write? I'm personally of the opinion that we should change the WASI layer to only allow at most one operation at a time if we can, meaning that each stream has a single pollable which indicates the status of the pending operation, if any. I know @pchickey you were thinking we should allow concurrent flush/write, though.

Maybe we should jump on a call for a chat about this?